src/Compiler/Mapper/Object/MapDiscriminatedObject.php

```diff
@@ -0,0 +1,160 @@
+
+        $statements = [
+            $builder->if($builder->not($builder->funcCall($builder->importFunction('is_array'), [$value])), [
+                $builder->throw(
+                    $builder->staticCall(
+                        $builder->importClass(MappingFailedException::class),
+                        'incorrectType',
+                        [$value, $path, $builder->val('array')],
+                    ),
+                ),
+            ]),
+        ];
+
+        $isDiscriminatorPresent = $builder->funcCall($builder->importFunction('array_key_exists'), [$builder->val($this->discriminatorFieldName), $value]);
+        $isDiscriminatorMissing = $builder->not($isDiscriminatorPresent);
+
+        $statements[] = $builder->if($isDiscriminatorMissing, [
+            $builder->throw(
+                $builder->staticCall(
+                    $builder->importClass(MappingFailedException::class),
+                    'missingKey',
+                    [$path, $this->discriminatorFieldName],
+                ),
+            ),
+        ]);
+
+        $discriminatorRawValue = $builder->arrayDimFetch($value, $builder->val($this->discriminatorFieldName));
+        $discriminatorPath = $builder->arrayImmutableAppend($path, $builder->val($this->discriminatorFieldName));
+        $discriminatorMapperMethodName = $builder->uniqMethodName('map' . ucfirst($this->discriminatorFieldName));
+        $discriminatorMapperMethod = $builder->mapperMethod($discriminatorMapperMethodName, new MapString())->makePrivate()->getNode();
+        $discriminatorMapperCall = $builder->methodCall($builder->var('this'), $discriminatorMapperMethodName, [$discriminatorRawValue, $discriminatorPath]);
+        $builder->addMethod($discriminatorMapperMethod);
+
+        $validMappingKeys = array_keys($this->subtypeMapping);
+        $isDiscriminatorValid = $builder->funcCall($builder->importFunction('in_array'), [$discriminatorRawValue, $builder->val($validMappingKeys), $builder->val(true)]);
+
+        $expectedDescription = $builder->concat(
+            'one of ',
+            $builder->funcCall($builder->importFunction('implode'), [
+                ', ',
+                $builder->val($validMappingKeys),
+            ]),
+        );
+
+        $statements[] = $builder->if($builder->not($isDiscriminatorValid), [
+            $builder->throw(
+                $builder->staticCall(
+                    $builder->importClass(MappingFailedException::class),
+                    'incorrectValue',
+                    [$discriminatorRawValue, $discriminatorPath, $expectedDescription],
+                ),
+            ),
+        ]);
+
+        $subtypeMatchArms = [];
+
+        foreach ($this->subtypeMapping as $key => $subtype) {
+            $mapperProviderMethodCall = $builder->methodCall($provider, 'get', [
+                $builder->classConstFetch($builder->importClass($subtype), 'class'),
```

Comment:
This is probably not working correctly with generics 🤔 

I should be passing the options from the parent.
The original version was accepting objectMappers directly from the factory - https://github.com/shipmonk-rnd/input-mapper/pull/68/commits/51bbc5c8ee0ab39fd45aeec6aa953ca2c686221a.

I can improve the MapperProvider or go back to the initial implementation.

What would you suggest @JanTvrdik?

Reply:
I think it's better to accept `array<string, MapperCompiler>`, but instead of `MapObject` you should pass any array of `DelegateMapperCompiler`.

Reply:
Replaced. 

Generics inside the subtype mapper properties works out of the box.
Generics on the outside, e.g.:
```php
#[Discriminator('type', mapping: [ 'a' => A::class])
class Base<T> {
    public mixed $val;
}

class Foo {
    /** Base<int> */
    Base $value;
}
```

cannot work due to the fact that it's not clear how the generic param from base-class should be mapped to the child (and it's an edge-case anyway)